File: src/test/java/com/tngtech/jgiven/LibraryTestUtil.java
package com.tngtech.jgiven;

import com.intellij.openapi.command.WriteCommandAction;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.roots.ModuleRootModificationUtil;
import com.intellij.openapi.roots.impl.libraries.ProjectLibraryTable;
import com.intellij.openapi.roots.libraries.Library;
import com.intellij.openapi.roots.libraries.LibraryTable;
import com.intellij.openapi.util.Ref;
import com.intellij.openapi.vfs.newvfs.impl.VfsRootAccess;
import com.intellij.testFramework.PsiTestUtil;
import com.tngtech.jgiven.annotation.ScenarioState;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class LibraryTestUtil {
    private final Module myModule;

    private List<Ref<Library>> references = new ArrayList<>();

    LibraryTestUtil(Module myModule) {
        this.myModule = myModule;
    }

    public LibraryTestUtil addJarContaining(Class<?> clazz) {
        String path = clazz.getProtectionDomain().getCodeSource().getLocation().getPath();
        addLibraryAt(path);
        return this;
    }

    public LibraryTestUtil addJGiven() {
        return addJarContaining(ScenarioState.class);
    }

    private void addLibraryAt(String path) {
        String[] parts = path.split("/");
        File file = new File(path);
        VfsRootAccess.allowRootAccess(parts[0]);
        String fileName = file.getName();

        ModuleRootModificationUtil.updateModel(myModule, model ->
                references.add(Ref.create(PsiTestUtil.addLibrary(myModule, model, fileName, file.getParent(), fileName))));
    }

    void removeLibraries() {
        WriteCommandAction.runWriteCommandAction(null, () -> {
            LibraryTable table = ProjectLibraryTable.getInstance(myModule.getProject());
            LibraryTable.ModifiableModel model = table.getModifiableModel();
            references.forEach(reference -> model.removeLibrary(reference.get()));
            model.commit();
        });
    }
}

File: src/test/java/com/tngtech/jgiven/LibraryTestUtil.kt
package com.tngtech.jgiven

import com.intellij.openapi.command.WriteCommandAction
import com.intellij.openapi.module.Module
import com.intellij.openapi.roots.ModuleRootModificationUtil
import com.intellij.openapi.roots.libraries.Library
import com.intellij.openapi.roots.libraries.LibraryTablesRegistrar
import com.intellij.openapi.util.Ref
import com.intellij.openapi.vfs.newvfs.impl.VfsRootAccess
import com.intellij.testFramework.PsiTestUtil
import com.tngtech.jgiven.annotation.ScenarioState
import java.io.File

class LibraryTestUtil constructor(private val myModule: Module) {

    private val references = ArrayList<Ref<Library>>()

    private fun addJarContaining(clazz: Class<*>): LibraryTestUtil {
        val path = getJarPath(clazz)
        addLibraryAt(path)
        return this
    }

    private fun getJarPath(clazz: Class<*>): String {
        val className = clazz.name.replace('.', '/') + ".class"
        val classPath = clazz.classLoader.getResource(className)?.toString()
                ?: throw IllegalArgumentException("Class $clazz not found")
        return if (classPath.startsWith("jar")) {
            classPath.substringAfter("jar:file:").substringBefore('!')
        } else {
            throw IllegalArgumentException("Class $clazz is not in a jar file")

        }
    }


    fun addJGiven(): LibraryTestUtil {
        try {
            return addJarContaining(ScenarioState::class.java)
        }catch(e: IllegalArgumentException){
            throw IllegalArgumentException("JGiven jar not found. Please add it to the classpath")
        }
    }

    private fun addLibraryAt(path: String) {
        val parts = path.split("/".toRegex()).filterNot { it.isEmpty() }.toTypedArray()
        val file = File(path)
        VfsRootAccess.allowRootAccess(myModule, parts[0])
        val fileName = file.name
        ModuleRootModificationUtil.updateModel(myModule) { model ->
            references.add(Ref.create(PsiTestUtil.addLibrary(
                    model, fileName, file.parent, fileName
            )))
        }
    }

    internal fun removeLibraries() {
        WriteCommandAction.runWriteCommandAction(null) {
            val table = LibraryTablesRegistrar.getInstance().getLibraryTable(myModule.project)
            val model = table.modifiableModel
            references.forEach { reference -> model.removeLibrary(reference.get()) }
            model.commit()
        }
    }
}